pkg/provision/storage/asyncStorage.go

```diff
@@ -0,0 +1,251 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/controllers/workspace/provision"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	devfileConstants "github.com/devfile/devworkspace-operator/pkg/library/constants"
+	"github.com/devfile/devworkspace-operator/pkg/provision/storage/asyncstorage"
+
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// The AsyncStorageProvisioner provisions one PVC per namespace and creates an ssh deployment that syncs data into that PVC.
+// Workspaces are provisioned with sync sidecars that sync data from the workspace to the async ssh deployment. All storage
+// attached to a workspace is emptyDir volumes.
+type AsyncStorageProvisioner struct{}
+
+var _ Provisioner = (*AsyncStorageProvisioner)(nil)
+
+func (*AsyncStorageProvisioner) NeedsStorage(workspace *dw.DevWorkspaceTemplateSpec) bool {
+	return needsStorage(workspace)
+}
+
+func (p *AsyncStorageProvisioner) ProvisionStorage(podAdditions *v1alpha1.PodAdditions, workspace *dw.DevWorkspace, clusterAPI provision.ClusterAPI) error {
+	numWorkspaces, _, err := p.getAsyncWorkspaceCount(clusterAPI)
+	if err != nil {
+		return err
+	}
+	// If there is more than one started workspace using async storage, then we fail starting additional ones
+	// Note we need to check phase so as to not accidentally fail an already-running workspace when a second one
+	// is created.
+	if numWorkspaces > 1 && workspace.Status.Phase != dw.WorkspaceStatusRunning {
```

Comment:
Part of the issue here is that I don't want to _actually_ restrict to 1 running workspace at a time -- the only reason we can't run multiple workspaces simultaneously is that cleanup becomes hard (we've no way to clean up the common PVC while it's mounted to the async server). 

The idea behind this snippet is to fail workspaces whenever we see multiple async workspaces in a namespace. To avoid failing an already-started workspace when a second one is created, we skip workspaces that are already Running, which _should_ (my logic might be wrong here) lead to the cases:
1. Start an async workspace -> starts fine
2. Start two+ async workspaces at same time -> both fail
3. Start a second async workspace when one is already running -> new workspace fails but running workspace is unaffected.

The second case is tricky, since it becomes a race to see which one starts first, and then we have to fail the second. It's safer to just fail hard as soon as possible. Ideally whoever is creating devworkspaces limits it to one per namespace for now (e.g. che dashboard). Since this check happens on reconciles, the only way an async workspace enters the `running` state (which requires a reconcile) is if it is the only async workspace on the cluster at the time that reconcile is triggered. 

The main gap I can see here is that I don't think we properly clean resources when a workspace fails. If a workspace fails after the deployment has been created, that deployment will stick around for now. 

Created issue https://github.com/devfile/devworkspace-operator/issues/338 to partially look at this.